oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2TokenIntrospectionAuthenticationProvider.java

```diff
@@ -0,0 +1,178 @@
+
+		TokenType tokenType = null;
+
+		try {
+			String tokenTypeHint = tokenIntrospectionAuthentication.getTokenTypeHint();
+			if (StringUtils.hasText(tokenTypeHint)) {
+				if (TokenType.REFRESH_TOKEN.getValue().equals(tokenTypeHint)) {
+					tokenType = TokenType.REFRESH_TOKEN;
+				} else if (TokenType.ACCESS_TOKEN.getValue().equals(tokenTypeHint)) {
+					tokenType = TokenType.ACCESS_TOKEN;
+				} else if (TokenType.ID_TOKEN.getValue().equals(tokenTypeHint)) {
+					tokenType = TokenType.ID_TOKEN;
+				} else {
+					throw new OAuth2AuthenticationException(new OAuth2Error(OAuth2ErrorCodes2.UNSUPPORTED_TOKEN_TYPE));
+				}
+			}
+
+			OAuth2Authorization authorization = this.authorizationService
+					.findByTokenWithHint(tokenIntrospectionAuthentication.getTokenValue(), Optional.ofNullable(tokenType))
+					.orElseThrow(() -> new IntrospectionTokenException("Token not found"));
+
+			if (!registeredClient.getId().equals(authorization.getRegisteredClientId())) {
+				throw new IntrospectionTokenException("Invalid client");
+			}
+
+			AbstractOAuth2Token token = authorization.getTokens().getToken(tokenIntrospectionAuthentication.getTokenValue())
+					.get();
+
+			if (authorization.getTokens().getTokenMetadata(token).isInvalidated()) {
+				throw new IntrospectionTokenException("Token has been invalidated");
+			}
+
+			token = parseJwt(token);
+
+			validateToken(token);
+
+			return new OAuth2TokenIntrospectionAuthenticationToken(clientPrincipal, registeredClient.getClientId(), token);
+		} catch (IntrospectionTokenException exception) {
+			return new OAuth2TokenIntrospectionAuthenticationToken(clientPrincipal, registeredClient.getClientId(), null);
+		}
+	}
+
+	@Override
+	public boolean supports(Class<?> authentication) {
+		return OAuth2TokenIntrospectionAuthenticationToken.class.isAssignableFrom(authentication);
+	}
+
+	private AbstractOAuth2Token parseJwt(AbstractOAuth2Token token) {
+		for (JwtDecoder jwtDecoder : this.jwtDecoders) {
+			try {
+				return jwtDecoder.decode(token.getTokenValue());
+			} catch (JwtException ex) {
+				// token might not be a JWT, or can't be processed with this decoder.
+			}
+		}
+		return token;
+	}
+
+	@SuppressWarnings("unchecked")
+	private <T extends AbstractOAuth2Token> void validateToken(T token) {
+		OAuth2TokenValidator<T> tokenValidator = (token instanceof Jwt) ? (OAuth2TokenValidator<T>) new JwtTimestampValidator()
+				: (OAuth2TokenValidator<T>) new TokenExpirationValidator();
+		OAuth2TokenValidatorResult result = tokenValidator.validate(token);
+		if (result.hasErrors()) {
+			String errorMessages = result.getErrors().stream().map(OAuth2Error::getErrorCode)
+					.collect(Collectors.joining(", ", "Invalid Token:", "."));
+			throw new IntrospectionTokenException(errorMessages);
+		}
+	}
+
+	/**
+	 * Exception that can be triggered when a token is found invalid.
+	 *
+	 * @author Gerardo Roza
+	 */
+	private static class IntrospectionTokenException extends RuntimeException {
```

Comment:
Remove this and instead use `OAuth2AuthenticationException`, e.g.

`throw new OAuth2AuthenticationException(new OAuth2Error(OAuth2ErrorCodes.INVALID_GRANT))`

Take a look at `OAuth2AuthorizationCodeAuthenticationProvider` and `OAuth2TokenRevocationAuthenticationProvider` for examples on how exceptions are thrown with associated OAuth 2.0 error codes.

Reply:
I am aware of the `OAuth2AuthenticationException` (and understand how this `AuthenticationException` fits in the documentation you suggested me to read here https://github.com/spring-projects-experimental/spring-authorization-server/pull/161#issuecomment-734426330), but note this is on purpose. Let me explain.

Simply put, **in most cases we don't want to retrieve an error response for this endpoint** (like for example an error response indicating the reason why the token is invalid) to avoid exposing any information about the server state, but instead, **we will simply retrieve an `{"active": false}` response**:

> If the protected resource uses OAuth 2.0 client credentials to
>   authenticate to the introspection endpoint and its credentials are
>   invalid, the authorization server responds with an HTTP 401
>   (Unauthorized) as described in Section 5.2 of OAuth 2.0 [RFC6749].
> ...
> **Note that a properly formed and authorized query** for an inactive or
>   otherwise invalid token (or a token the protected resource is not
>   allowed to know about) **is not considered an error response by this**
>   **specification.  In these cases, the authorization server MUST instead**
>   **respond with an introspection response with the "active" field set to**
>   **"false"** as described in Section 2.2.
>   ...
>   Note that **to avoid disclosing too**
>   **much of the authorization server's state** to a third party, the
>   authorization server **SHOULD NOT include any additional information**
>   **about an inactive token, including why the token is inactive.**

So, here:

- we're covering the first case thanks to the `getAuthenticatedClientElseThrowInvalidClient` functionality which does throw an `OAuth2AuthenticationException` when suitable
- and for the rest of the cases, I use this internal `IntrospectionTokenException`, **without leaving it to bubble up**; so we can **stop processing the token at any point of the operation** where we find out the token is not valid and t**he AuthenticationProvider will retrieve an authenticated `OAuth2TokenIntrospectionAuthenticationToken` (bc we're in fact authenticated), but in a way that will result in a `{"active": false }` response**.

If I used a `OAuth2AuthenticationException` for these cases, I would have to catch it, analyze somehow the reason (if it was in fact because the request was unauthorized, or if it was because the token is invalid), and as you can imagine this will be unnecessarily complex.

Of course, if you want to suggest a better approach, I'd love to hear it, as this would help me improve and use better solutions :)

Reply:
I would still remove `IntrospectionTokenException` as it's not used anyway other than catching it and returning a `OAuth2TokenIntrospectionAuthenticationToken`. Creating and throwing exceptions is costly so we should avoid it unless we use it.

Instead, perform a fail-fast and return `OAuth2TokenIntrospectionAuthenticationToken` at the point validation fails. You can see that [OAuth2TokenRevocationAuthenticationProvider](https://github.com/spring-projects-experimental/spring-authorization-server/blob/ee1b46b9a627504ed2d0fdaf7308b11fba23fb98/oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2TokenRevocationAuthenticationProvider.java#L68) does this.

Reply:
Sounds good, thanks for the clarification Joe, this is now resolved 👍